Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+                    Point point = line.End;
+                    pointsChain.Add(point);
+                    lastPoint = point;
+                }
+                else
+                {
+                    Point pt1 = line.Start;
+                    Point pt2 = line.End;
+
+                    if (lastPoint.Distance(pt1) < lastPoint.Distance(pt2) && !((Math.Abs(lastPoint.Distance(cameraLocation) - radius) < tolerance) && (Math.Abs(pt2.Distance(cameraLocation) - radius) < tolerance)) && i > 0)
+                    {
+                        Polyline obst = lineObstacledictionary[i].Values.First();
+                        Polyline lastObst = lineObstacledictionary[i - 1].Values.First();
+                        if (!lastPoint.IsEqual(cameraLocation) && lastObst == obst)
+                        {
+                            pointsChain.Add(pt2);
+                            pointsChain.Add(pt1);
+                            lastPoint = pt1;
+                            continue;
+                        }
+                        pointsChain.Add(pt1);
+                        pointsChain.Add(pt2);
+                        lastPoint = pt2;
+                    }
+                    else
+                    {
+                        pointsChain.Add(pt2);
+                        pointsChain.Add(pt1);
+                        lastPoint = pt1;
+                    }
+                }
+            }
+            pointsChain = pointsChain.CullDuplicates(tolerance);
+            pointsChain.Add(cameraLocation);
+
+            return pointsChain;
+        }
+
+        /***************************************************/
+
+        [Description("Create PolyCurve that represents camera field of view.")]
+        [Input("pointsChain", "Points chain of the camera view cone.")]
+        [Input("cameraLocation", "Location of the camera device.")]
+        [Input("radius", "Radius of the camera view cone.")]
+        [Output("viewCone", "PolyCurve that represents camera field of view.")]
+        private static PolyCurve CameraViewPolyCurve(this List<Point> pointsChain, Point cameraLocation, double radius, Arc coneArc, double tolerance)
+        {
+            List<ICurve> curves = new List<ICurve>();
+            for (int i = 1; i < pointsChain.Count; i++)
+            {
+                Point pt1 = pointsChain[i - 1];
+                Point pt2 = pointsChain[i];
+
+                if ((Math.Abs(pt1.Distance(cameraLocation) - radius) < tolerance) && (Math.Abs(pt2.Distance(cameraLocation) - radius) < tolerance))
+                {
+                    double p1Param = coneArc.ParameterAtPoint(pt1, tolerance);
+                    double p2Param = coneArc.ParameterAtPoint(pt2, tolerance);
+                    double p3Param = (p1Param + p2Param) / 2;
+                    Point pt3 = coneArc.PointAtParameter(p3Param);
+                    Arc newArc = BH.Engine.Geometry.Create.Arc(pt1, pt3, pt2, tolerance);
+                    curves.Add(newArc);
```

Comment:
I think this may be a bit heavy - instead, couldn't you create the arc from pt1 to pt2 knowing the coordinate system of the original arc? just wondering if that could not be sped up. if too much work, please skip, i do not think this method contributes to the total execution time heavily.

Reply:
Moved to issue #2890.